cf-oscore/src/main/java/org/eclipse/californium/oscore/ObjectSecurityLayer.java

```diff
@@ -349,4 +382,105 @@ private static boolean shouldProtectRequest(Request request) {
 	private static boolean isProtected(Message message) {
 		return message.getOptions().getOscore() != null;
 	}
+
+	/**
+	 * Check if a message being sent exceeds the MAX_UNFRAGMENTED_SIZE and is
+	 * not using inner block-wise. If so it should not be sent.
+	 * 
+	 * @param message the CoAP message
+	 * @param maxUnfragmentedSize the MAX_UNFRAGMENTED_SIZE value
+	 * 
+	 * @return if the message exceeds the MAX_UNFRAGMENTED_SIZE
+	 */
+	private boolean outgoingExceedsMaxUnfragSize(Message message, boolean outerBlockwise,
+			int maxUnfragmentedSize) {
+
+		boolean usesInnerBlockwise = (message.getOptions().hasBlock1() == true
+				|| message.getOptions().hasBlock2() == true) && outerBlockwise == false;
+
+		if (message.getPayloadSize() > maxUnfragmentedSize && usesInnerBlockwise == false) {
+			return true;
+		} else {
+			return false;
+		}
+
+	}
+
+	/**
+	 * Check if the cumulative payload size for an incoming block-wise transfer
+	 * exceeds the MAX_UNFRAGMENTED_SIZE and is not using inner block-wise. If
+	 * so reception should be aborted. This method is only called when using
+	 * outer block-wise so no need to check that explicitly.
+	 * 
+	 * @param message the CoAP message
+	 * @param exchange the exchange
+	 * @param ctxDb the context database used
+	 * 
+	 * @return if the cumulative payload exceeds the MAX_UNFRAGMENTED_SIZE
+	 */
+	private boolean incomingExceedsMaxUnfragSize(Message message, Exchange exchange, OSCoreCtxDB ctxDb) {
+
+		int cumulativeSize;
+		OSCoreCtx ctx = null;
+		if (message instanceof Request) {
+			ctx = ctxDb.getContext(getRid(message.getOptions().getOscore()));
+			cumulativeSize = blockwiseTransferInfo.getReceivedBodySize(exchange, (Request) message);
+			cumulativeSize += message.getPayloadSize(); // Count current payload
+		} else {
+			ctx = ctxDb.getContextByToken(message.getToken());
```

Comment:
And fro responses the Token is used to lookup the ctx, which contains the max size?

Reply:
Yes, since I know which context the corresponding request was sent using, I can look up that same context using the Token for the response.